File: limiters_cuda.py
import math
from numba import cuda
import numba

@cuda.jit(device=True, inline=True)
def venkat_limiter(qtilde_shared, q_gpu, maxminq, dq, nbhs, conn, x, y, min_dist, idx, VL_CONST, shared, delx, dely, gamma):
    epsi = VL_CONST * min_dist[idx]
    epsi = math.pow(epsi,3)
    q = q_gpu[idx]
    qtilde_shared[cuda.threadIdx.x + cuda.blockDim.x * 0] = 1
    qtilde_shared[cuda.threadIdx.x + cuda.blockDim.x * 1] = 1
    qtilde_shared[cuda.threadIdx.x + cuda.blockDim.x * 2] = 1
    qtilde_shared[cuda.threadIdx.x + cuda.blockDim.x * 3] = 1

    for i in range(4):
        del_neg = q_gpu[idx][i] - (0.5 * (delx * dq[idx][0][i] + dely * dq[idx][1][i])) - q[i]
        if math.fabs(del_neg) > 1e-5:
            if del_neg > 0:
                del_pos = maxminq[idx][0][i] - q[i]
            elif del_neg < 0:
                del_pos = maxminq[idx][1][i] - q[i]

            num = (del_pos*del_pos) + (epsi*epsi)
            num = num*del_neg + 2.0*del_neg*del_neg*del_pos

            den = del_pos*del_pos + 2.0*del_neg*del_neg
            den = den + del_neg*del_pos + epsi*epsi
            den = den*del_neg

            temp = num/den

            if temp < 1:
                qtilde_shared[cuda.threadIdx.x + cuda.blockDim.x * i] = temp

    for i in range(4):
        qtilde_shared[cuda.threadIdx.x + cuda.blockDim.x * i] = q_gpu[idx][i] - 0.5 * (qtilde_shared[cuda.threadIdx.x + cuda.blockDim.x * i] * (delx * dq[idx][0][i] + dely * dq[idx][1][i]))

    beta = - qtilde_shared[cuda.threadIdx.x + cuda.blockDim.x * 3] * 0.5

    temp = 0.5/beta

    u1 = qtilde_shared[cuda.threadIdx.x + cuda.blockDim.x * 1] * temp
    u2 = qtilde_shared[cuda.threadIdx.x + cuda.blockDim.x * 2] * temp

    temp1 = qtilde_shared[cuda.threadIdx.x] + beta * ( u1 * u1 + u2 * u2 )
    temp2 = temp1 - (math.log(beta)/(gamma-1))
    rho = math.exp(temp2)
    pr = rho * temp

    shared[cuda.threadIdx.x + cuda.blockDim.x * 4] = u1
    shared[cuda.threadIdx.x + cuda.blockDim.x * 5] = u2
    shared[cuda.threadIdx.x + cuda.blockDim.x * 6] = rho
    shared[cuda.threadIdx.x + cuda.blockDim.x * 7] = pr

@cuda.jit(device=True, inline=True)
def maximum(globaldata, idx, i, maxval, itm):
    maxval[i] = globaldata[idx]['q'][i]
    if maxval[i] < globaldata[itm]['q'][i]:
        maxval[i] = globaldata[itm]['q'][i]

@cuda.jit(device=True, inline=True)
def minimum(globaldata, idx, i, minval, itm):
    minval[i] = globaldata[idx]['q'][i]
    if minval[i] > globaldata[itm]['q'][i]:
        minval[i] = globaldata[itm]['q'][i]

@cuda.jit(device=True, inline=True)
def smallest_dist(globaldata, idx, min_dist):
    min_dist[0] = 10000

    for itm in globaldata[idx]['conn'][:globaldata[idx]['nbhs']]:
        dx = globaldata[idx]['x'] - globaldata[itm]['x']
        dy = globaldata[idx]['y'] - globaldata[itm]['y']
        ds = math.sqrt(dx * dx + dy * dy)
        
        if ds < min_dist[0]:
            min_dist[0] = ds

# def max_q_values(globaldata, idx, temp):

#     maxq = globaldata[idx]['q']

#     for itm in globaldata[idx]['conn']:
#         currq = globaldata[itm]['q']
#         for i in range(4):
#             if maxq[i] < currq[i]:
#                 maxq[i] = currq[i]
    
#     return maxq

# def min_q_values(globaldata, idx, temp):
#     minq = globaldata[idx]['q']

#     for itm in globaldata[idx]['conn']:
#         currq = globaldata[itm]['q']
#         for i in range(4):
#             if minq[i] > currq[i]:
#                 minq[i] = currq[i]
    
#     return minq

File: JULIA_MESHFREE_CUDA/limiters_cuda.jl
function venkat_limiter_kernel(gpuGlobalDataFixedPoint, gpuGlobalDataRest, idx, gpuConfigData, delx, dely, shared, thread_idx)
    VL_CONST = gpuConfigData[8]
    # @cuprintf("Type is %s", typeof(VL_CONST))
    epsigh = VL_CONST * gpuGlobalDataFixedPoint[idx].short_distance
    epsi = CUDAnative.pow(epsigh, 3.0)
    del_pos = 0.0
    del_neg = 0.0

    for i in 1:4
        q = gpuGlobalDataRest[8+i, idx]
        del_neg = gpuGlobalDataRest[8+i, idx] - 0.5*(delx * gpuGlobalDataRest[12+i, idx] + dely * gpuGlobalDataRest[16+i, idx]) - q
        if abs(del_neg) <= 1e-5
            shared[thread_idx + i] = 1.0
        elseif abs(del_neg) > 1e-5
            if del_neg > 0.0
                # maximum(globaldata, idx, i, max_q)
                del_pos = gpuGlobalDataRest[20+i, idx] - q
            elseif del_neg < 0.0
                # minimum(globaldata, idx, i, min_q)
                del_pos = gpuGlobalDataRest[24+i, idx] - q
            end
            num = (del_pos*del_pos) + (epsi*epsi)
            num = (num*del_neg) + 2 * (del_neg*del_neg*del_pos)

            den = (del_pos*del_pos) + (2 *del_neg*del_neg)
            den = den + (del_neg*del_pos) + (epsi*epsi)
            den = den*del_neg

            temp = num/den
            if temp > 1.0
                temp = 1.0
            end
            shared[thread_idx + i] = temp
        end
    end
    return nothing
end


@inline function qtilde_to_primitive_kernel(qtilde, gpuConfigData, shared, thread_idx)
    idx = (blockIdx().x - 1) * blockDim().x + threadIdx().x
    # # gamma = gpuConfigData[15]
    beta = -qtilde[4]*0.5
    temp = 0.5/beta
    u1 = qtilde[2]*temp
    u2 = qtilde[3]*temp

    temp2 = qtilde[1] + beta*(u1*u1 + u2*u2) - (CUDAnative.log(beta)/(gpuConfigData[15]-1))
    rho = CUDAnative.exp(temp2)
    shared[thread_idx + 5] = u1
    shared[thread_idx + 6] = u2
    shared[thread_idx + 7] = rho
    shared[thread_idx + 8] = rho*temp
    return nothing
end


# function max_q_values_kernel(gpuGlobalDataCommon, idx, maxq)
#     maxq = (
#                 gpuGlobalDataRest[9, idx],
#                 gpuGlobalDataRest[10, idx],
#                 gpuGlobalDataRest[11, idx],
#                 gpuGlobalDataRest[12, idx]
#             )
#     for iter in 9:28
#         conn = Int(gpuGlobalDataCommon[iter, idx])
#         if conn == 0.0
#             break
#         end
#         # currq = globaldata[itm].q
#         if maxq[1] < gpuGlobalDataRest[9, conn]
#             maxq = (gpuGlobalDataRest[9, conn], maxq[2], maxq[3], maxq[4])
#         end
#         if maxq[2] < gpuGlobalDataRest[10, conn]
#             maxq = (maxq[1], gpuGlobalDataRest[10, conn], maxq[3], maxq[4])
#         end
#         if maxq[3] < gpuGlobalDataRest[11, conn]
#             maxq = (maxq[1], maxq[2], gpuGlobalDataRest[11, conn], maxq[4])
#         end
#         if maxq[4] < gpuGlobalDataRest[12, conn]
#             maxq = (maxq[1], maxq[2], maxq[3], gpuGlobalDataRest[12, conn])
#         end
#     end
#     return nothing
# end

# function min_q_values_kernel(gpuGlobalDataCommon, idx, minq)
#     minq = (
#                 gpuGlobalDataRest[9, idx],
#                 gpuGlobalDataRest[10, idx],
#                 gpuGlobalDataRest[11, idx],
#                 gpuGlobalDataRest[12, idx]
#             )
#     for iter in 9:28
#         conn = Int(gpuGlobalDataCommon[iter, idx])
#         if conn == 0.0
#             break
#         end
#         # currq = globaldata[itm].q
#         if minq[1] > gpuGlobalDataRest[9, conn]
#             minq = (gpuGlobalDataRest[9, conn], minq[2], minq[3], minq[4])
#         end
#         if minq[2] > gpuGlobalDataRest[10, conn]
#             minq = (minq[1], gpuGlobalDataRest[10, conn], minq[3], minq[4])
#         end
#         if minq[3] > gpuGlobalDataRest[11, conn]
#             minq = (minq[1], minq[2], gpuGlobalDataRest[11, conn], minq[4])
#         end
#         if minq[4] > gpuGlobalDataRest[12, conn]
#             minq = (minq[1], minq[2], minq[3], gpuGlobalDataRest[12, conn])
#         end
#     end
#     return nothing
# end

# @inline function smallest_dist(globaldata, idx::Int64)
#     min_dist = 1000.0
#     for itm in globaldata[idx].conn
#         ds = hypot(globaldata[idx].x - globaldata[itm].x, globaldata[idx].y - globaldata[itm].y)
#         if ds < min_dist
#             min_dist = ds
#         end
#     end
#     globaldata[idx].short_distance = min_dist
# end
